docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,132 @@
+---
+title: bmac-reboot-hosts-on-iso-change
+authors:
+  - "@mhrivnak"
+  - "@avishayt"
+  - "@flaper87"
+creation-date: 2021-06-30
+last-updated: 2021-07-01
+---
+
+# BMAC reboots only those hosts impacted by an ISO change
+
+## Summary
+
+When an InfraEnv's ISO gets regenerated based on a change such as adding or
+changing an NMStateConfig resource, there may be only a subset of that
+InfraEnv's Agents that are affected by the change. Rather than reboot all of
+them with the newly-generated ISO, only those hosts that are affected by the
+new or modified NMStateConfig should be rebooted.
+
+## Motivation
+
+If there are 500 NMStateConfigs and 500 Agents all related to the same
+InfraEnv, and then someone adds the 501st NMStateConfig, we do not want to
+reboot all 500 Agents. Rebooting is an expensive and time-consuming operation.
+We need a way to determine which Agents or hosts should be rebooted in response
+to an NMStateConfig.
+
+### Goals
+
+* Avoid rebooting un-provisioned BareMetalHosts when it is not needed.
+* Leave the door open for other platforms (include None/boot-it-yourself) to also avoid host reboots that are not needed.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+The assisted-service will track whether a particular Agent was booted using an
+old InfraEnv or NMStateConfig. It will ensure that any Agent that was booted
+using obsolete artifacts will have:
+* a label indicating that it was booted from obsolete artifacts
+* an array in its Status of references to the resources that are obsolete
+
+### assisted-service InfraEnv controller
+
+The InfraEnv ISO URL will change when the ISO contents change. This will make
+it obvious to an API user that they have booted a host from an obsolete ISO,
+because the URLs won't match. A side benefit of ensuring that the ISO URL
+uniquely identifies the resource's contents, is that pull-through caches will
+have an easy time understanding when an file is no longer relevant.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and compared against
+the current resource generations to determine if they are obsolete.
+
+### assisted-service Agent controller
+
+When the Agent resource gets created, its status should include the
+above-described reference details for:
+* The InfraEnv
+* The NMStateConfig it utilized during its boot
```

Comment:
Is that a new logic that need to be added to the agent? when it boots it will just send the inventory no?
How will it know what is the relevant nmstate cr?

Reply:
Yes this requires new logic. Today the ISO includes data from many NMStateConfigs, and the Agent at runtime picks the one relevant to it (by matching a MAC address). We'll need to enhance that task so that the data also includes a reference to the k8s resource.

Reply:
Is there a case to support only layer 3 configuration?

Reply:
What do you mean exactly? The basic use case is to identify a layer 2 device by MAC address and then apply layer 3 configuration to it.

There is some layer 2 configuration we can do, such as bonding, but that's optional.

Reply:
In https://docs.google.com/document/d/1PEOfM3-f_JUtR0rjZ74W_j3_8HGbKDrgrU3yVUTdhZQ/edit#heading=h.9cdnoxyarh8q
 there is a section related to `Add L3 connectivity check` i wasn't sure if it mean that we don't use l2 there. maybe it's not relevant

Reply:
Checking L3 connectivity would just be a validation. I suspect that's based on the prior cases where we've seen a host configured without a default route.